README.md
# Dorado sensitivity and exposure time calculator

Dorado is a proposed space mission for ultraviolet follow-up of gravitational
wave events. This repository contains a simple sensitivity and exposure time
calculator for Dorado.

This package can estimate the signal to noise, exposure time, or limiting
magnitude for an astronomical source with a given spectrum using the [CCD
signal to noise equation]. It models the following noise contributions:

*   Zodiacal light
*   Airglow (geocoronal emission)
*   Standard CCD noise (shot noise, read noise, dark current)

## Installation

To install with [Pip]:

    $ pip install dorado-sensitivity

## Examples

For examples, see the [Jupyter notebook].

## Dependencies

*   [Astropy]
*   [Synphot] for combining bandpasses and source spectra
*   [PyYAML] for reading [ECSV] data files

[CCD signal to noise equation]: https://hst-docs.stsci.edu/stisihb/chapter-6-exposure-time-calculations/6-4-computing-exposure-times
[Pip]: https://pip.pypa.io
[Astropy]: https://www.astropy.org
[Synphot]: https://synphot.readthedocs.io/
[PyYAML]: https://pyyaml.org/
[ECSV]: https://github.com/astropy/astropy-APEs/blob/master/APE6.rst
[Jupyter notebook]: https://github.com/nasa/dorado-sensitivity/blob/master/example.ipynb


CHANGES.md
# Changelog

## Version 0.3.0 (2021-03-18)

-   The ``get_limmag`` function now accepts ``synphot.SourceSpectrum``
    instance, rather than a spectral model class, for more consistency with the
    other methods and for more flexibility.

## Version 0.2.0 (2021-03-03)

-   Update the instrument characteristics with the values in the Dorado concept
    study report. The effective area and the spot size both increased slightly.

-   Remove the ``bandpass`` argument for convenience, because Dorado has a
    single filter.

## Version 0.1.0 (2021-01-11)

-   First public release.


dorado/sensitivity/tests/test_sensitivity.py
#
# Copyright © 2020 United States Government as represented by the Administrator
# of the National Aeronautics and Space Administration. No copyright is claimed
# in the United States under Title 17, U.S. Code. All Other Rights Reserved.
#
# SPDX-License-Identifier: NASA-1.3
#
from astropy.coordinates import SkyCoord
from astropy.time import Time
from astropy import units as u
from hypothesis import given, settings
from hypothesis.strategies import booleans, floats, just
from synphot import ConstFlux1D, SourceSpectrum
from pytest import approx

from .. import get_exptime, get_limmag, get_snr


@given(
    snr=floats(1, 1e9),
    exptime=floats(1, 1e9),
    ra=just(0),
    dec=just(90),
    time=just(Time('2020-01-01')),
    night=booleans())
@settings(deadline=None)
def test_round_trip_snr_limmag(snr, exptime, ra, dec, time, night):
    """Test round trip: get_snr(get_limmag(...))"""
    kwargs = dict(
        exptime=exptime * u.s,
        coord=SkyCoord(ra * u.deg, dec * u.deg),
        time=time, night=night)

    limmag = get_limmag(
        SourceSpectrum(ConstFlux1D, amplitude=0*u.ABmag), snr=snr, **kwargs)
    snr_2 = get_snr(SourceSpectrum(ConstFlux1D, amplitude=limmag), **kwargs)
    assert snr_2 == approx(snr)


@given(
    snr=floats(1, 1e9),
    mag=floats(15, 22),
    ra=just(0),
    dec=just(90),
    time=just(Time('2020-01-01')),
    night=booleans())
@settings(deadline=None)
def test_round_trip_snr_exptime(snr, mag, ra, dec, time, night):
    """Test round trip: get_snr(get_limmag(...))"""
    source = SourceSpectrum(ConstFlux1D, amplitude=mag * u.ABmag)
    kwargs = dict(
        coord=SkyCoord(ra * u.deg, dec * u.deg),
        night=night, time=time)

    exptime = get_exptime(source, snr=snr, **kwargs)
    snr_2 = get_snr(source, exptime=exptime, **kwargs)
    assert snr_2 == approx(snr)


dorado/sensitivity/constants.py
#
# Copyright © 2020 United States Government as represented by the Administrator
# of the National Aeronautics and Space Administration. No copyright is claimed
# in the United States under Title 17, U.S. Code. All Other Rights Reserved.
#
# SPDX-License-Identifier: NASA-1.3
#
"""Scalar constants."""
from astropy import units as u
import numpy as np

AREA = 100.0 * u.cm**2
"""Fiducial collecting area.

Notes
-----
This is a fictitious area that is only used for converting effective area to
synphot dimensionless units. It is only equal to the actual collecting area of
the telescope to within an order of magnitude.
"""

PLATE_SCALE = 25.0 * u.arcsec * u.pix**-1
"""Plate scale.

Notes
-----
Assuing a fixed focal length and 2x2 binning.
"""

NPIX = np.pi * 0.89**2
"""Effective number of pixels in a circular aperture.

Notes
-----
26.8 um = 22.3" = 0.89 pix
"""

APERTURE_CORRECTION = 0.7
"""Aperture correction.

70% of the signal from a point source falls within the above circular aperture.
"""

DARK_NOISE = 4. * 30. * 0.124 * u.hour**-1
"""
Dark noise rate.

Notes
-----
From JPL spreadsheet; 4x for binning; 30x for MDF.
"""

READ_NOISE = 5.0
"""
Read noise.

Notes
-----
From CBE in proposal.
"""


dorado/sensitivity/bandpasses.py
#
# Copyright © 2020 United States Government as represented by the Administrator
# of the National Aeronautics and Space Administration. No copyright is claimed
# in the United States under Title 17, U.S. Code. All Other Rights Reserved.
#
# SPDX-License-Identifier: NASA-1.3
#
"""Bandpass filter curves."""
from astropy.table import QTable
from synphot import Empirical1D, SpectralElement

from importlib import resources

from . import constants
from . import data

__all__ = ('NUV_D',)


def _get_bandpass(name):
    with resources.path(data, f'{name}_effective_area.ecsv') as p:
        t = QTable.read(p)
    return SpectralElement(Empirical1D,
                           points=t['wavelength'],
                           lookup_table=t['effective_area'] / constants.AREA)


NUV_D = _get_bandpass('nuv_d')


dorado/sensitivity/backgrounds.py
#
# Copyright © 2020 United States Government as represented by the Administrator
# of the National Aeronautics and Space Administration. No copyright is claimed
# in the United States under Title 17, U.S. Code. All Other Rights Reserved.
#
# SPDX-License-Identifier: NASA-1.3
#
"""Sky backgrounds."""
from importlib import resources

from astropy.coordinates import GeocentricTrueEcliptic, get_sun, SkyCoord
from astropy.table import QTable
from astropy import units as u
import numpy as np
from scipy.interpolate import interp2d
from synphot import Empirical1D, GaussianFlux1D, SourceSpectrum

from . import constants
from . import data

__all__ = ('get_zodiacal_light', 'get_airglow')


def _get_zodi_angular_interp():
    # Zodiacal light angular dependence from Table 16 of Leinert et al. (2017)
    # https://doi.org/10.1051/aas:1998105.
    with resources.path(data, 'leinert_zodi.txt') as p:
        table = np.loadtxt(p)
    lat = table[0, 1:]
    lon = table[1:, 0]
    s10 = table[1:, 1:]

    # The table only extends up to a latitude of 75°. According to the paper,
    # "Towards the ecliptic pole, the brightness as given above is 60 ± 3 S10."
    lat = np.append(lat, 90)
    s10 = np.append(s10, np.tile(60.0, (len(lon), 1)), axis=1)

    # The table is in units of S10: the number of 10th magnitude stars per
    # square degree. Convert to magnitude per square arcsecond.
    sb = 10 - 2.5 * np.log10(s10 / 60**4)

    return interp2d(lat, lon, sb, bounds_error=True)


_zodi_angular_dependence = _get_zodi_angular_interp()


# Read zodiacal light spectrum
with resources.path(data, 'stis_zodi_high.ecsv') as p:
    table = QTable.read(p)
_stis_zodi_high = SourceSpectrum(
    Empirical1D,
    points=table['wavelength'],
    lookup_table=table['surface_brightness'] * u.arcsec**2)
del table


def _get_zodi_angular_dependence(coord, time):
    obj = SkyCoord(coord).transform_to(GeocentricTrueEcliptic(equinox=time))
    sun = get_sun(time).transform_to(GeocentricTrueEcliptic(equinox=time))

    # Wrap angles and look up in table
    lat = np.abs(obj.lat.deg)
    lon = np.abs((obj.lon - sun.lon).wrap_at(180 * u.deg).deg)
    result = _zodi_angular_dependence(lat, lon)

    # When interp2d encounters infinities, it returns nan. Fix that up here.
    result = np.where(np.isnan(result), -np.inf, result)

    # Fix up shape
    if obj.isscalar:
        result = result.item()

    return result - _zodi_angular_dependence(0, 180).item()


def get_zodiacal_light(coord, time):
    """Get the zodiacal light spectrum incident on one pixel.

    Estimate the zodiacal light spectrum based on the angular dependence
    (Table 6.2) and wavelength (Table 6.4) from the STIS Instrument Manual.

    Parameters
    ----------
    coord : astropy.coordinates.SkyCoord
        The coordinates of the object under observation. If the coordinates do
        not specify a distance, then the object is assumed to be a fixed star
        at infinite distance for the purpose of calculating its helioecliptic
        position.
    time : astropy.time.Time
        The time of the observation.

    Returns
    -------
    synphot.SourceSpectrum
        The zodiacal light spectrum, normalized to one pixel.

    References
    ----------
    https://hst-docs.stsci.edu/stisihb/chapter-6-exposure-time-calculations/6-5-detector-and-sky-backgrounds

    """
    scale = u.mag(1).to_physical(_get_zodi_angular_dependence(coord, time))
    scale *= ((constants.PLATE_SCALE * u.pix)**2).to_value(u.arcsec**2)
    return _stis_zodi_high * scale


def get_airglow(night):
    """Get the airglow spectrum incident on one pixel.

    Estimate the zodiacal light spectrum based on the [O II] geocoronal
    emission line (Table 6.5) in the STIS Instrument Manual.

    Parameters
    ----------
    night : bool
        Use the "low" value if True, or the "average" value if False.

    Returns
    -------
    synphot.SourceSpectrum
        The airglow spectrum, normalized to one pixel.

    References
    ----------
    https://hst-docs.stsci.edu/stisihb/chapter-6-exposure-time-calculations/6-5-detector-and-sky-backgrounds

    """
    flux = 1.5e-17 if night else 1.5e-15
    flux *= u.erg * u.s**-1 * u.cm**-2 * u.arcsec**-2
    flux *= (constants.PLATE_SCALE * u.pix)**2
    return SourceSpectrum(GaussianFlux1D,
                          mean=2471 * u.angstrom,
                          fwhm=0.023 * u.angstrom,
                          total_flux=flux)


dorado/sensitivity/__init__.py
#
# Copyright © 2020 United States Government as represented by the Administrator
# of the National Aeronautics and Space Administration. No copyright is claimed
# in the United States under Title 17, U.S. Code. All Other Rights Reserved.
#
# SPDX-License-Identifier: NASA-1.3
#
"""Dorado sensitivity calculator"""
from astropy.stats import signal_to_noise_oir_ccd
from astropy import units as u
import numpy as np
from synphot.exceptions import SynphotError
from synphot import Observation

from . import backgrounds
from . import bandpasses
from . import constants

__all__ = ('get_snr', 'get_limmag', 'get_exptime')


def _get_count_rate(source_spectrum):
    observation = Observation(source_spectrum, bandpasses.NUV_D)
    try:
        return observation.countrate(constants.AREA) / u.ct
    except SynphotError as e:
        if e.args[0] == 'Integrated flux is infinite':
            return np.inf * u.s**-1
        else:
            raise


def get_snr(source_spectrum, *, exptime, coord, time, night):
    """Calculate the SNR of an observation of a point source with Dorado.

    Parameters
    ----------
    source_spectrum : synphot.SourceSpectrum
        The spectrum of the source.
    exptime : astropy.units.Quantity
        The exposure time
    coord : astropy.coordinates.SkyCoord
        The coordinates of the source, for calculating zodiacal light
    time : astropy.time.Time
        The time of the observation, for calculating zodiacal light
    night : bool
        Whether the observation occurs on the day or night side of the Earth,
        for estimating airglow

    Returns
    -------
    float
        The signal to noise ratio
    """
    return signal_to_noise_oir_ccd(
        exptime,
        constants.APERTURE_CORRECTION * _get_count_rate(source_spectrum),
        (
            _get_count_rate(backgrounds.get_zodiacal_light(coord, time)) +
            _get_count_rate(backgrounds.get_airglow(night))
        ),
        constants.DARK_NOISE,
        constants.READ_NOISE,
        constants.NPIX
    ).to_value(u.dimensionless_unscaled)


def _amp_for_signal_to_noise_oir_ccd(
        snr, t, source_eps, sky_eps, dark_eps, rd, npix, gain=1.0):
    """Inverse of astropy.stats.signal_to_noise_oir_ccd."""
    signal = t * source_eps * gain
    # noise squared without signal shot noise term
    snr2 = np.square(snr)
    noise2 = t * (npix * (sky_eps * gain + dark_eps)) + npix * np.square(rd)
    return 0.5 * snr2 / signal * (1 + np.sqrt(1 + 4 * noise2 / snr2))


def get_limmag(source_spectrum, *, snr, exptime, coord, time, night):
    """Get the limiting magnitude for a given SNR.

    Parameters
    ----------
    source_spectrum : synphot.SourceSpectrum
        The spectrum of the source.
    snr : float
        The desired SNR.
    exptime : astropy.units.Quantity
        The exposure time
    coord : astropy.coordinates.SkyCoord
        The coordinates of the source, for calculating zodiacal light
    time : astropy.time.Time
        The time of the observation, for calculating zodiacal light
    night : bool
        Whether the observation occurs on the day or night side of the Earth,
        for estimating airglow

    Returns
    -------
    astropy.units.Quantity
        The AB magnitude of the source
    """
    mag0 = Observation(source_spectrum, bandpasses.NUV_D).effstim(
        u.ABmag, area=constants.AREA)

    result = _amp_for_signal_to_noise_oir_ccd(
        snr,
        exptime,
        constants.APERTURE_CORRECTION * _get_count_rate(source_spectrum),
        (
            _get_count_rate(backgrounds.get_zodiacal_light(coord, time)) +
            _get_count_rate(backgrounds.get_airglow(night))
        ),
        constants.DARK_NOISE,
        constants.READ_NOISE,
        constants.NPIX
    ).to_value(u.dimensionless_unscaled)

    return -2.5 * np.log10(result) * u.mag + mag0


def _exptime_for_signal_to_noise_oir_ccd(
        snr, source_eps, sky_eps, dark_eps, rd, npix, gain=1.0):
    """Inverse of astropy.stats.signal_to_noise_oir_ccd."""
    c1 = source_eps * gain
    c2 = npix * (sky_eps * gain + dark_eps)
    c3 = npix * np.square(rd)
    x = 1 + c2 / c1
    snr2 = np.square(snr)
    return 0.5 * snr2 / c1 * (x + np.sqrt(np.square(x) + 4 * c3 / snr2))


def get_exptime(source_spectrum, *, snr, coord, time, night):
    """Calculate the SNR of an observation of a point source with Dorado.

    Parameters
    ----------
    source_spectrum : synphot.SourceSpectrum
        The spectrum of the source.
    snr : float
        The signal to noise ratio
    coord : astropy.coordinates.SkyCoord
        The coordinates of the source, for calculating zodiacal light
    time : astropy.time.Time
        The time of the observation, for calculating zodiacal light
    night : bool
        Whether the observation occurs on the day or night side of the Earth,
        for estimating airglow

    Returns
    -------
    astropy.units.Quantity
        The exposure time
    """
    return _exptime_for_signal_to_noise_oir_ccd(
        snr,
        constants.APERTURE_CORRECTION * _get_count_rate(source_spectrum),
        (
            _get_count_rate(backgrounds.get_zodiacal_light(coord, time)) +
            _get_count_rate(backgrounds.get_airglow(night))
        ),
        constants.DARK_NOISE,
        constants.READ_NOISE,
        constants.NPIX
    )
